impl/src/main/java/com/okta/sdk/impl/oauth2/AccessTokenRetrieverServiceImpl.java

```diff
@@ -0,0 +1,258 @@
+
+    private static final String TOKEN_URI  = "/oauth2/v1/token";
+
+    private ClientConfiguration tokenClientConfiguration;
+    private OAuth2TokenClient tokenClient;
+
+    public AccessTokenRetrieverServiceImpl(ClientConfiguration apiClientConfiguration) {
+        Assert.notNull(apiClientConfiguration, "apiClientConfiguration must not be null.");
+        ClientConfiguration tokenClientConfig = constructTokenClientConfig(apiClientConfiguration);
+        this.tokenClient = new OAuth2TokenClient(tokenClientConfig);
+        this.tokenClientConfiguration = tokenClientConfig;
+    }
+
+    public AccessTokenRetrieverServiceImpl(ClientConfiguration apiClientConfiguration, OAuth2TokenClient tokenClient) {
+        Assert.notNull(apiClientConfiguration, "apiClientConfiguration must not be null.");
+        Assert.notNull(tokenClient, "tokenClient must not be null.");
+        this.tokenClient = tokenClient;
+        this.tokenClientConfiguration = constructTokenClientConfig(apiClientConfiguration);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OAuth2AccessToken getOAuth2AccessToken() throws IOException, InvalidKeyException, OAuth2TokenRetrieverException {
+        log.debug("Attempting to get OAuth2 access token for client id {} from {}",
+            tokenClientConfiguration.getClientId(), tokenClientConfiguration.getBaseUrl() + TOKEN_URI);
+
+        String signedJwt = createSignedJWT();
+        String scope = String.join(" ", tokenClientConfiguration.getScopes());
+
+        try {
+            ExtensibleResource accessTokenResponse = tokenClient.http()
+                .addHeaderParameter("Accept", MediaType.APPLICATION_JSON_VALUE)
+                .addHeaderParameter("Content-Type", MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+                .addQueryParameter("grant_type", "client_credentials")
+                .addQueryParameter("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+                .addQueryParameter("client_assertion", signedJwt)
+                .addQueryParameter("scope", scope)
+                .post(TOKEN_URI, ExtensibleResource.class);
+
+            OAuth2AccessToken oAuth2AccessToken = new OAuth2AccessToken();
+            oAuth2AccessToken.setTokenType(accessTokenResponse.getString(OAuth2AccessToken.TOKEN_TYPE_KEY));
+            oAuth2AccessToken.setExpiresIn(accessTokenResponse.getInteger(OAuth2AccessToken.EXPIRES_IN_KEY));
+            oAuth2AccessToken.setAccessToken(accessTokenResponse.getString(OAuth2AccessToken.ACCESS_TOKEN_KEY));
+            oAuth2AccessToken.setScope(accessTokenResponse.getString(OAuth2AccessToken.SCOPE_KEY));
+
+            log.debug("Got OAuth2 access token for client id {} from {}",
+                tokenClientConfiguration.getClientId(), tokenClientConfiguration.getBaseUrl() + TOKEN_URI);
+
+            return oAuth2AccessToken;
+        } catch (Exception e) {
+            if (e instanceof ResourceException) {
```

Comment:
nit: you should probably add a catch block for `ResourceException` instead of doing an if/else  with a`instanceof`